File: Julia/Token.jl
#:----------------------------------------------------------------------------
#:
#:  Copyright © 2023 Steffen Liersch
#:  https://www.steffen-liersch.de/
#:
#:----------------------------------------------------------------------------

struct Token
  asnumber::Union{Number,Nothing}
  asstring::String
  Token(value::Number) = new(value, string(value))
  Token(value::String) = (x = parsetoken(value); new(x[2], x[1]))
end

function parsetoken(value::String)::Tuple{String,Union{Float64,Nothing}}
  v = tryparse(Float64, value)
  if v !== nothing
    return value, v
  end
  s = uppercase(value)
  if s == "E"
    return s, MathConstants.e
  elseif s == "PI"
    return s, MathConstants.pi
  else
    return s, nothing
  end
end

tostring(token::Token) = token.asstring

File: TypeScript/token.ts
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

export { Token };

class Token
{
  readonly asNumber?: number;
  readonly asString: string;

  constructor(value: number | string)
  {
    if(typeof value == 'number')
    {
      this.asNumber = value;
      this.asString = '' + value;
    }
    else
    {
      let v = Number(value);
      if(!Number.isNaN(v))
      {
        this.asNumber = v;
        this.asString = value;
      }
      else
      {
        switch(value.toUpperCase())
        {
          case 'E':
            this.asNumber = Math.E;
            this.asString = 'E';
            break;

          case 'PI':
            this.asNumber = Math.PI;
            this.asString = 'PI';
            break;

          default:
            this.asString = value;
            break;
        }
      }
    }
  }

  toString(): string { return this.asString; }
}
